text/0000-local-snapshot-file-format/0000-local-snapshot-file-format.md

```diff
@@ -0,0 +1,147 @@
++ Feature name: `local_snapshot_file_format`
++ Start date: 2020-08-25
++ RFC PR: [iotaledger/protocol-rfcs#0000](https://github.com/iotaledger/protocol-rfcs/pull/0000)
+
+# Summary
+
+This RFC defines a file format for local snapshots which is compatible with Chrysalis Phase 2.
+
+# Motivation
+
+Nodes create local snapshots to produce ledger representations at a point in time of a given milestone to be able to:
+* Start up from a recent milestone instead of having to synchronize from the genesis transaction.
+* Delete old transaction data below a given milestone.
+
+Current node implementations use a [local snapshot file format](https://github.com/iotaledger/iri-ls-sa-merger/tree/351020d3b5e342b6e9a41f2868575ab7ff8c251c#generating-an-export-file-from-a-localsnapshots-db) which only works with account based ledgers. For Chrysalis Phase 2, this file format has to be assimilated to support a UTXO based ledger.
+
+# Detailed design
+
+Since a UTXO based ledger is much larger in size, this RFC proposes two formats for snapshot files: 
+* A `full` format which represents a complete ledger state.
+* A `delta` format which only contains diffs (consumed and spent outputs) of milestones from a given milestone index onwards.
+
+This separation allows nodes to swiftly create new delta snapshot files, which then can be distributed with a companion full snapshot file to reconstruct a recent state.
+
+Unlike the current format, these new formats do not include spent addresses since this information is no longer held by nodes.
+
+### Formats
+
+> All types are serialized in little-endian
+
+#### Full Ledger State
+
+A full ledger snapshot file contains the UTXOs (`outputs` section) of a node's latest solid milestone (`ledger_milestone_index`). The `diffs` contain the diffs to rollback the `outputs` state to regain the ledger state of the snapshot milestone at (`seps_milestone_index`).
+
+![](https://i.imgur.com/e6WuufK.png)
+
+While the node producing such a full ledger state snapshot could theoretically pre-compute the actual snapshot milestone state, this is deferred to the consumer of the data to speed up local snapshot creation.
+
+A full ledger state local snapshot is denoted by the type byte `0`:
+
+```
+version<byte>
+type<byte> = 0
+timestamp<uint64>
+network_id<uint64>
+seps_milestone_index<uint64>
+ledger_milestone_index<uint64>
+seps_count<uint64>
+outputs_count<uint64>
+diffs_count<uint64>
+seps<array[seps_count]>:
+	sep<array[32]>
+outputs<array[outputs_count]>:
+	transaction_hash<array[32]>
+	output_index<uint16>
+	address:
+		address_type<byte>
+		wots_address<array[49]>
+		||
+		address_type<byte>
+		ed25519_address<array[32]>
+	value<uint64>
+diffs<array[diffs_count]>:
```

Comment:
There is no specific ordering of diffs you enforce here. You can have a diff for milestone X followed by a diff for milestone X+100. For integrity and validation sake it would be nicer to have an `array[ledger_milestone_index - seps_milestone_index]` of an array of diffs between milestone X and X+1, sorted by X, starting from `seps_milestone_index` and ending on `ledger_milestone_index`.